x86/HVM: return all ones on wrong-sized reads of system device I/O ports

So far the value presented to the guest remained uninitialized.

This is XSA-121.

Signed-off-by: Jan Beulich <jbeulich@suse.com>
Acked-by: Ian Campbell <ian.campbell@citrix.com>

Index: xen-4.1.4/xen/arch/x86/hvm/i8254.c
===================================================================
--- xen-4.1.4.orig/xen/arch/x86/hvm/i8254.c	2013-01-22 18:41:53.000000000 +0200
+++ xen-4.1.4/xen/arch/x86/hvm/i8254.c	2015-03-04 17:51:06.000000000 +0200
@@ -475,6 +475,7 @@
     if ( bytes != 1 )
     {
         gdprintk(XENLOG_WARNING, "PIT bad access\n");
+        *val = ~0;
         return X86EMUL_OKAY;
     }
 
Index: xen-4.1.4/xen/arch/x86/hvm/pmtimer.c
===================================================================
--- xen-4.1.4.orig/xen/arch/x86/hvm/pmtimer.c	2013-01-22 18:41:53.000000000 +0200
+++ xen-4.1.4/xen/arch/x86/hvm/pmtimer.c	2015-03-04 17:51:06.000000000 +0200
@@ -213,6 +213,7 @@
     if ( bytes != 4 )
     {
         gdprintk(XENLOG_WARNING, "HVM_PMT bad access\n");
+        *val = ~0;
         return X86EMUL_OKAY;
     }
     
Index: xen-4.1.4/xen/arch/x86/hvm/rtc.c
===================================================================
--- xen-4.1.4.orig/xen/arch/x86/hvm/rtc.c	2013-01-22 18:41:53.000000000 +0200
+++ xen-4.1.4/xen/arch/x86/hvm/rtc.c	2015-03-04 17:51:06.000000000 +0200
@@ -408,7 +408,8 @@
 
     if ( bytes != 1 )
     {
-        gdprintk(XENLOG_WARNING, "HVM_RTC bas access\n");
+        gdprintk(XENLOG_WARNING, "HVM_RTC bad access\n");
+        *val = ~0;
         return X86EMUL_OKAY;
     }
     
Index: xen-4.1.4/xen/arch/x86/hvm/vpic.c
===================================================================
--- xen-4.1.4.orig/xen/arch/x86/hvm/vpic.c	2013-01-22 18:41:53.000000000 +0200
+++ xen-4.1.4/xen/arch/x86/hvm/vpic.c	2015-03-04 17:51:06.000000000 +0200
@@ -324,6 +324,7 @@
     if ( bytes != 1 )
     {
         gdprintk(XENLOG_WARNING, "PIC_IO bad access size %d\n", bytes);
+        *val = ~0;
         return X86EMUL_OKAY;
     }
 
